### PR TITLE
8317678: Fix up hashCode() for ZipFile.Source.Key

### PR DESCRIPTION
**Pull request by coffeys**
Fix up java.util.zip.ZipFile$Source hashCode() impl so that duplicate Source objects aren't created for the same zip file.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8317678](https://bugs.openjdk.org/browse/JDK-8317678): Fix up hashCode() for ZipFile.Source.Key (**Bug** - P3)

### Reviewers
 * [Lance Andersen](https://openjdk.org/census#lancea) (@LanceAndersen - **Reviewer**) ⚠️ Review applies to [ff6c7d70](https://git.openjdk.org/jdk/pull/16115/files/ff6c7d7050125f327fdbfba4431e74471886db9c)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [3635e542](https://git.openjdk.org/jdk/pull/16115/files/3635e5424c5c385396896c5b0d4a0b49abf235f1)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16115/head:pull/16115` \
`$ git checkout pull/16115`

Update a local copy of the PR: \
`$ git checkout pull/16115` \
`$ git pull https://git.openjdk.org/jdk.git pull/16115/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16115`

View PR using the GUI difftool: \
`$ git pr show -t 16115`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16115.diff">https://git.openjdk.org/jdk/pull/16115.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16115#issuecomment-1754706555)

### CODE REVIEW

File: src/java.base/share/classes/java/util/zip/ZipFile.java
Comment (AlanBateman) [inline src/java.base/share/classes/java/util/zip/ZipFile.java:1424]: Mapping to RuntimeException doesn't look right here, instead need to use PrivilegedExceptionAction here and changing the Key constructor to throw IOException.

File: src/java.base/share/classes/java/util/zip/ZipFile.java
Comment (pavelrappo) [inline src/java.base/share/classes/java/util/zip/ZipFile.java:1444]: Consider this trivial simplification:
```suggestion
                return Long.hashCode(t)
                        + (fk != null ? fk.hashCode() : file.hashCode());
```

**Comment by AlanBateman** [comment]
Can you look at the equals method too, need to decide if the checking of lastModifiedTime should be removed.

File: src/java.base/share/classes/java/util/zip/ZipFile.java
Comment (liach) [inline src/java.base/share/classes/java/util/zip/ZipFile.java:1418]:
```suggestion
            private static File getCanonicalFile(File file) {
```
To avoid this-escape.

**Comment by coffeys** [comment]
Thanks for the comments to date. I think the latest commit should satisfy requests made.

for `Key.equals(Object)` and `Key.hashCode()` - I've changed the impl so that attrs.lastModifiedTime() is only used in the case where `attrs.fileKey()` returns null. Is that in the direction you were suggesting Alan ?

File: src/java.base/share/classes/java/util/zip/ZipFile.java
Comment (AlanBateman) [inline src/java.base/share/classes/java/util/zip/ZipFile.java:1426]: The IOException to throw is PAE's cause, meaning you need to unwrap.

File: src/java.base/share/classes/java/util/zip/ZipFile.java
Comment (AlanBateman) [inline src/java.base/share/classes/java/util/zip/ZipFile.java:1423]: I think you'll have to cleanly separate the SM and no-SM case because of the exception handling, then you can use `PrivilegedExceptionAction<File> pae = file::getCanonicalFile`.

File: src/java.base/share/classes/java/util/zip/ZipFile.java
Comment (AlanBateman) [inline src/java.base/share/classes/java/util/zip/ZipFile.java:1451]: The question in the previous comment is why is the last modified time used here, I think you have to dig into why it is used by equals.

**Comment by coffeys** [comment]
`lastModifiedTime()` was used in the native code and brought across when native code was removed in JDK-8145260

I thought we might be able to drop it for the case where we have a unique file key but that won't cover the case where an existing zip file is opened and modified. Re-introduced the lastModifiedTime checks as a result. We need to be able to detect such scenarios, otherwise new ZipFile constructs would continue to use a stale <Key, Source> mapping (if such Keys still existed)

I've updated the test case to junit format and beefed it up with some extra scenarios to cover the above.

**Review by LanceAndersen — APPROVED** [review]
Thank you for driving this issue Sean.

Overall I think it looks good.

Please see my minor comment below which you can handle prior to pushing the PR

File: test/jdk/java/util/zip/ZipFile/ZipSourceCache.java
Comment (jaikiran) [inline test/jdk/java/util/zip/ZipFile/ZipSourceCache.java:59]: Hello Sean, this looks like a typo to me. I think it should have been `ZIPFILE_NAME` instead of `ZIPENTRY_NAME`.

**Comment by jaikiran** [comment]
I had a look at the linked JBS and the new jtreg test case in this PR. Is my understanding correct that the issue that we are trying to solve here is the case where the cache would end up having 2 separate entries for the same jar file, if one `JarFile` was constructed using an absolute path and the other `JarFile` (for the same underlying file on the filesystem) was constructed using a relative path?

The mention of fixing hashCode() makes me wonder if there is/was some other issue that was triggered which lead to creation of that JBS issue.

Coming to the proposed implementation in this PR, I see that we first rely on `BasicFileAttributes.fileKey()` and if it is available for a given `File` instance, we store that `File` instance as a component of the `Key`. Looking at the javadoc of `BasicFileAttributes.fileKey()` it states:

```
* Returns an object that uniquely identifies the given file, or {@code
* null} if a file key is not available. On some platforms or file systems
* it is possible to use an identifier, or a combination of identifiers to
* uniquely identify a file. Such identifiers are important for operations
* such as file tree traversal in file systems that support <a
* href="../package-summary.html#links">symbolic links</a> or file systems
* that allow a file to be an entry in more than one directory. On UNIX file
* systems, for example, the <em>device ID</em> and <em>inode</em> are
* commonly used for such purposes.
```
It's unclear to me (and I haven't yet read up on inode management in Unix) if 2 different symbolic links pointing to the same target (jar) file on the filesystem would have 2 different/unique `fileKey()`. If they do, then do you think that even with the current proposed solution of using `fileKey()`, we would still end up with 2 different entries in this cache, each with a different `Key`? Would that then bring us back to the original issue?

As for the use of `File.getCanonicalFile()` in the absence of a `fileKey()`, there used to be performance considerations with the use of `File.getCanonicalFile()` (and `File.getCanonicalPath()`) to the extent that there used to be a cache involved. Recent versions of Java have removed that cache because it was no longer necessary as detailed in https://bugs.openjdk.org/browse/JDK-8301001. I don't have historical knowledge behind the motivation of that cache (JDK-8301001 does provide some necessary history), however JDK-8301001 also says that one of the reasons why the cache is no longer needed is because:

> The work in JDK 9 to remove canonicalization from FilePermission (JDK-8164705) mostly eliminated the need for this caching.

Would introducing this `File.getCanonicalFile()` in this `JarFile` area now reintroduce performance considerations (for startup?) on filesystems that don't return a `fileKey()`? I see that the implementation of `File.getCanonicalFile()` now ends up being a native call.

File: src/java.base/share/classes/java/util/zip/ZipFile.java
Comment (AlanBateman) [inline src/java.base/share/classes/java/util/zip/ZipFile.java:1423]: The exception handling isn't right, you should be re-throwing pae.getException when it's an IOException, this should do what you want
```
        PrivilegedExceptionAction<File> pa = file::getCanonicalFile;
        try {
            return AccessController.doPrivileged(pa);
        } catch (PrivilegedActionException pae) {
            Throwable cause = pae.getCause();
            if (cause instanceof IOException ioe) {
                throw ioe;
            } else {
                throw new IOException(cause);
            }
        }
```

File: src/java.base/share/classes/java/util/zip/ZipFile.java
Comment (AlanBateman) [inline src/java.base/share/classes/java/util/zip/ZipFile.java:1430]: ZipFile cannot tolerate a zip file being changed while in use. I realise you don't want to change this part of the code but I think it is desperately needs a comment in both hashCode and equals to explain why the last modified time is included.

**Comment by coffeys** [comment]
thanks for the reviews and comments to date. I've pushed new changes which should address them. 

Good catch on the ZIPENTRY_NAME typo @jaikiran 
I've run some tests also and it looks like the fileKey() returns the same object/reference - even if symbolic links are used.

you raise an interesting point about `getCanonicalFile()` performance though. I've updated an existing benchmark[1] to check and performance for fileystems that don't support fileKey() (windows) has regressed[2]. Performance is much better on systems that support fileKey() (unix) --  I'll look at it further but perhaps I'll only target the optimizations for the filesystems that support filekey() for now.

[1]
```
diff --git a/test/micro/org/openjdk/bench/java/util/zip/ZipFileOpen.java b/test/micro/org/openjdk/bench/java/util/zip/ZipFileOpen.java
index ffbbc3d245f..34449c14e32 100644
--- a/test/micro/org/openjdk/bench/java/util/zip/ZipFileOpen.java
+++ b/test/micro/org/openjdk/bench/java/util/zip/ZipFileOpen.java
@@ -29,6 +29,7 @@
 import java.io.FileOutputStream;
 import java.io.IOException;
 import java.nio.file.Files;
+import java.nio.file.Path;
 import java.util.concurrent.TimeUnit;
 import java.util.zip.ZipEntry;
 import java.util.zip.ZipFile;
@@ -50,6 +51,7 @@ public class ZipFileOpen {
     private int size;

     public File zipFile;
+    public File relativePathFile;

     @Setup(Level.Trial)
     public void beforeRun() throws IOException {
@@ -74,6 +76,8 @@ public void beforeRun() throws IOException {
             }
         }
         zipFile = tempFile;
+        relativePathFile = Path.of(System.getProperty("user.dir"))
+                                .relativize(zipFile.toPath()).toFile();
     }

     @Benchmark
@@ -90,4 +94,13 @@ public ZipFile openCloseZipFile() throws Exception {
         zf.close();
         return zf;
     }
+
+    @Benchmark
+    public void openCloseZipFilex2() throws Exception {
+        // Ensure that we only create ZipFile.Source.Key entry per unique zip file
+        ZipFile zf = new ZipFile(zipFile);
+        ZipFile zf2 = new ZipFile(relativePathFile);
+        zf.close();
+        zf2.close();
+    }
 }
```
[2]

jdk-tip-no-patch:  (Windows)

```
Benchmark                       (size)  Mode  Cnt       Score        Error  Units
ZipFileOpen.openCloseZipFile       512  avgt   15  319543.232 ±  46112.481  ns/op
ZipFileOpen.openCloseZipFile      1024  avgt   15  463719.381 ± 139409.316  ns/op
ZipFileOpen.openCloseZipFilex2     512  avgt   15  750838.029 ± 190726.572  ns/op
ZipFileOpen.openCloseZipFilex2    1024  avgt   15  875174.035 ± 140971.036  ns/op

jdk-with-new-patch
Benchmark                       (size)  Mode  Cnt        Score        Error  Units
ZipFileOpen.openCloseZipFile       512  avgt   15   603053.335 ±  50514.066  ns/op
ZipFileOpen.openCloseZipFile      1024  avgt   15   710960.542 ±  46343.033  ns/op
ZipFileOpen.openCloseZipFilex2     512  avgt   15  1258566.172 ± 216230.090  ns/op
ZipFileOpen.openCloseZipFilex2    1024  avgt   15  1260380.961 ± 164216.628  ns/op
```

[3]

```
without patch: (Linux)

Benchmark                       (size)  Mode  Cnt       Score       Error  Units
ZipFileOpen.openCloseZipFile       512  avgt   15   75378.890 ?   642.370  ns/op
ZipFileOpen.openCloseZipFile      1024  avgt   15  150814.293 ? 29801.424  ns/op
ZipFileOpen.openCloseZipFilex2     512  avgt   15  168104.286 ? 41505.778  ns/op
ZipFileOpen.openCloseZipFilex2    1024  avgt   15  299282.488 ? 44496.777  ns/op
Finished running test 'micro:java.util.zip.ZipFileOpen'


with patch:

Benchmark                       (size)  Mode  Cnt       Score       Error  Units
ZipFileOpen.openCloseZipFile       512  avgt   15   75400.408 ?   408.981  ns/op
ZipFileOpen.openCloseZipFile      1024  avgt   15  146213.650 ? 15509.167  ns/op
ZipFileOpen.openCloseZipFilex2     512  avgt   15   83194.552 ? 16361.844  ns/op
ZipFileOpen.openCloseZipFilex2    1024  avgt   15  153956.028 ? 30550.595  ns/op
```

**Comment by coffeys** [comment]
Removed the `getCanonicalFile `calls. The patch now optimizes the case where fileKey() is supported. Updated the ZipFileOpen.java benchmark test case also. before and after benchmark results for linux and windows posted below

```
linux:
without patch:

Benchmark                       (size)  Mode  Cnt       Score       Error  Units
ZipFileOpen.openCloseZipFile       512  avgt   15   75378.890 ?   642.370  ns/op
ZipFileOpen.openCloseZipFile      1024  avgt   15  150814.293 ? 29801.424  ns/op
ZipFileOpen.openCloseZipFilex2     512  avgt   15  168104.286 ? 41505.778  ns/op
ZipFileOpen.openCloseZipFilex2    1024  avgt   15  299282.488 ? 44496.777  ns/op
Finished running test 'micro:java.util.zip.ZipFileOpen'

with patch
Benchmark                       (size)  Mode  Cnt       Score       Error  Units
ZipFileOpen.openCloseZipFile       512  avgt   15   82715.346 ? 21700.950  ns/op
ZipFileOpen.openCloseZipFile      1024  avgt   15  142253.235 ?  1054.553  ns/op
ZipFileOpen.openCloseZipFilex2     512  avgt   15   77238.320 ?   194.585  ns/op
ZipFileOpen.openCloseZipFilex2    1024  avgt   15  161828.411 ? 45486.890  ns/op
Finished running test 'micro:java.util.zip.ZipFileOpen'

windows:
without patch:
ZipFileOpen.openCloseZipFile       512  avgt   15  319543.232 ±  46112.481  ns/op
ZipFileOpen.openCloseZipFile      1024  avgt   15  463719.381 ± 139409.316  ns/op
ZipFileOpen.openCloseZipFilex2     512  avgt   15  750838.029 ± 190726.572  ns/op
ZipFileOpen.openCloseZipFilex2    1024  avgt   15  875174.035 ± 140971.036  ns/op

with patch:
Benchmark                       (size)  Mode  Cnt       Score        Error  Units
ZipFileOpen.openCloseZipFile       512  avgt   15  400393.718 ± 161890.652  ns/op
ZipFileOpen.openCloseZipFile      1024  avgt   15  380922.933 ±  21776.006  ns/op
ZipFileOpen.openCloseZipFilex2     512  avgt   15  623589.801 ±  38110.510  ns/op
ZipFileOpen.openCloseZipFilex2    1024  avgt   15  774654.146 ±  50413.718  ns/op
```

File: test/jdk/java/util/zip/ZipFile/ZipSourceCache.java
Comment (LanceAndersen) [inline test/jdk/java/util/zip/ZipFile/ZipSourceCache.java:103]: Perhaps put the expected message in a constant in the event we need to change it at some point it is easier to find

**Review by LanceAndersen — APPROVED** [review]
Latest updates look good.

**Review by LanceAndersen — APPROVED** [review]
Thank you for the last set up updates Sean

File: src/java.base/share/classes/java/util/zip/ZipFile.java
Comment (AlanBateman) [inline src/java.base/share/classes/java/util/zip/ZipFile.java:1415]: I see you've added a comment to the checking of the last modified time but I think the comment is a bit confusing as there isn't any detection of a changed zip file in the hashCode/equals methods.  I think you want to say that the lastModifiedTime is included in the hash code to allow for cases where the same file is opened more than once and where it has been modified in the mean-time.

**Review by LanceAndersen — APPROVED** [review]

File: src/java.base/share/classes/java/util/zip/ZipFile.java
Comment (AlanBateman) [inline src/java.base/share/classes/java/util/zip/ZipFile.java:1431]: The code change to the hashCode method is fine but I think the comments added to both hashCode + equals are still a bit confusing because they hint that the last modified time is used to detect if a zip file has been modified. What would you think about dropping these comments and instead put a class file comment on Key to say that it represents a key to a zip file. It key is based on the file key if available, or the canonical path if the file key is not available. The key is also based on the file's last modified time to allow for cases where a zip file is re-opened after it has been modified.

**Comment by jaikiran** [comment]
Hello Sean, thank you for the updates. The fix now really boils down to fixing the `hashCode()` implementation of `java.util.zip.ZipFile$Source$Key` so that it honours the `equals()` and `hashCode()` contract which mandates that the `hashCode()` returned by instances which are `equals()` must be the same.

The change you have in the `Key` class now looks good to me.

Coming to the newly introduced `ZipSourceCache.java` test case, I went through it. What's being done in that test looks OK to me, but I felt that it's a bit too complex for what I think we should be testing here - asserting that `hashCode()` returned by `Key` instances are same if `Key` instances are `equals()`.

I decided to attempt a different way to test this and this is what the test looks like:

```java
/*
 * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

/* @test
 * @bug 8317678
 * @modules java.base/java.util.zip:open
 * @summary Fix up hashCode() for ZipFile.Source.Key
 * @run junit/othervm ZipSourceCache
 */

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.lang.reflect.Constructor;
import java.lang.reflect.Method;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.nio.file.FileSystem;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.attribute.BasicFileAttributes;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

public class ZipSourceCache {
    private static final String TEST_FILE_NAME = "foobar.zip";

    @BeforeAll
    public static void beforeAll() throws Exception {
        // create the file to run our test
        try (final FileOutputStream fos = new FileOutputStream(TEST_FILE_NAME)) {
            // no need to write anything to the file, just the presence of the file
            // is enough
        }
    }

    @AfterAll
    public static void afterAll() throws Exception {
        Files.deleteIfExists(Path.of(TEST_FILE_NAME));
    }

    /*
     * Verifies that instances of java.util.zip.ZipFile$Source$Key which are "equal()"
     * return back the same "hashCode()"
     */
    @Test
    public void testKeyEqualsHashCode() throws Exception {
        // create an instance of java.util.zip.ZipCoder for UTF-8 charset
        final Class<?> zipCoderClass = Class.forName("java.util.zip.ZipCoder");
        final Method m = zipCoderClass.getDeclaredMethod("get", Charset.class);
        m.setAccessible(true);
        final Object utf8ZipCoder = m.invoke(null, StandardCharsets.UTF_8);


        // get the java.util.zip.ZipFile$Source$Key class
        final Class<?> keyClass = Class.forName("java.util.zip.ZipFile$Source$Key");
        final Constructor<?> ctr = keyClass.getConstructor(File.class, BasicFileAttributes.class,
                zipCoderClass);
        ctr.setAccessible(true);

        final File relFile = new File(TEST_FILE_NAME);
        final File absFile = new File(TEST_FILE_NAME).getAbsoluteFile();
        System.out.println("Running test on " + keyClass.getName() + " with relative File path "
                + relFile.getPath() + " and absolute File path " + absFile.getPath());
        // create an instance of java.util.zip.ZipFile$Source$Key passing it a relative file path
        // to "foobar.zip"
        final Object relKey = ctr.newInstance(relFile, getAttrs(relFile), utf8ZipCoder);
        System.out.println("relKey, hashCode = " + relKey.hashCode());

        // create an instance of java.util.zip.ZipFile$Source$Key passing it an absolute file path
        // to "foobar.zip"
        final Object absKey = ctr.newInstance(absFile, getAttrs(absFile), utf8ZipCoder);
        System.out.println("absKey, hashCode = " + absKey.hashCode());
        // these 2 Key instances, one created with a relative File path and other with a absolute
        // File path, to the same file, are expected to be "equal()"
        Assertions.assertTrue(absKey.equals(relKey), "Key instances were expected to be equal");
        // verify that the instances which are "equal()" return the same "hashCode()"
        Assertions.assertEquals(relKey.hashCode(), absKey.hashCode(), "Differing hashCode() for" +
                " Keys that were equal");
    }

    private static BasicFileAttributes getAttrs(final File file) throws IOException {
        final FileSystem fs = file.toPath().getFileSystem();
        return Files.readAttributes(fs.getPath(file.getPath()),
                BasicFileAttributes.class);
    }
}
```
All it does is asserts the `hashCode()` values of `Key` instances created out of a relative file and absolute file pointing to the same underlying file. This test consistently fails without your fix for the `Key` class and passes with your fix. I haven't run it on Windows, but I think it shouldn't be any different.

Do you think we could simplify the `ZipSourceCache.java` test in this PR to instead use this alternate approach or some form of it?

**Review by AlanBateman — APPROVED** [review]
The update to ZipFile.Key hashCode and the Key comments look good. I don't have time to study the test but I think Lance and Jai are looking at that.

**Comment by jaikiran** [comment]
> ZipFile behaviour in the scenario where a file is re-opened and edited
> while references to the old ZipFile exist. A "invalid LOC header" needs to be 
> thrown in cases where an old zip file mapping is still in use.

This PR doesn't do any source changes to introduce this behaviour. It instead just adds a test to assert this existing behaviour. So that's a good thing and testing this behaviour sounds fine to me.

>   the behaviour of incorrectly adding extra <Key, Source> mappings to our 
> internal map has gone undetected for a few years. 
> We've no test coverage in that area. The current test closely monitors 
> numbers in that map now and allows for behavioural differences between 
> Filesystems that support and don't support fileKey(). I think that protects us 
> from future changes we might make in this area also.

What you propose here sounds fine. The only part that I think will need a bit more experiments is to run something like a test-repeat of 50 or more to make sure that the number of jars opened when this test is running  doesn't get impacted by any unforeseen loading of classes or such (from the test framework for example) which could end up opening some more jar files which can then cause the shared static map to have a different count of jars than this test would expect. This is already a `/othervm` test so intereference is already reduced, but it would still be good to see if there's any scope of intermittent failures, by runinng a test-repeat job.

> I think your test will fail on windows. The rel and abs file Keys are still treated as different there (proposal to use getCanonicalPath seems too costly for performance)

You are right, I missed that part where the `equals()` may return false on Windows. So that part can't actually be asserted. The `hashCode()` too then needs to be asserted only if the instances are `equals()`.

> While your suggested test case captures the core issue at hand, I think it misses a few important points that I captured in the current test.

Thank you for the additional details explaining the rationale behind this test. What you propose sounds fine to me then. I've a few minor nits about this test, which I'll add inline.

File: test/jdk/java/util/zip/ZipFile/ZipSourceCache.java
Comment (jaikiran) [inline test/jdk/java/util/zip/ZipFile/ZipSourceCache.java:80]: Do you think changing this to:
```java
new ZipFile(relativeFile)
```
would be better and more clear about the intent? A few lines below, we use `new ZipFile(absoluteFile)` which I think is good because it makes it clear what is being done.

File: test/jdk/java/util/zip/ZipFile/ZipSourceCache.java
Comment (jaikiran) [inline test/jdk/java/util/zip/ZipFile/ZipSourceCache.java:108]: I think adding a else block which prints that this part of the testing was skipped might provide some visibility on which parts of this test were run, if/when debugging this test on CI environments.

File: test/jdk/java/util/zip/ZipFile/ZipSourceCache.java
Comment (jaikiran) [inline test/jdk/java/util/zip/ZipFile/ZipSourceCache.java:98]: Same comment as a few lines above, maybe we should use `new ZipFile(relativeFile)` here to make the intention clear?

**Comment by jaikiran** [comment]
Finally, now that this PR only fixes a bug in the hashCode() and doesn't introduce any performance changing characteristics to the code, do you think the micro benchmark test update being proposed in this PR is needed? I don't have too much knowledge of `jmh` and it's a bit unclear to me if the introduction of this new `@Benchmark` method would interefere with the numbers of the existing `@Benchmark` method. I can spend some time on reading up `jmh` and then reviewing this micro benchmark if you think adding that benchmark is useful in context of this PR.

**Comment by coffeys** [comment]
Thanks for the review Jai. I've edited the test to use` ZipFile z = new ZipFile(relativeFile)` in both places. 

Regards the JMH benchmark, note that there is a significant performance gain for some environments with this patch. 2X improvement for filesystems that support fileKey() and deal with opening of a zip file twice (or more) but with different variants of path name. The benchmark update was useful for highlighting this. The existing ` src = files.get(key);` ZipFile code call is key to this improvement. Now that hashcode() is corrected, a new <Key, Source> pairing is avoided for the same zip file.

**Comment by jaikiran** [comment]
Thank you for the updates, Sean. The test `ZipSourceCache` looks good to me in its current form. I have no other review comments except for the jmh benchmark. I'll approve this PR in its current form.

As for the jmh benchmark, the existing benchmark method has some informative inline comment explaining what it does. Perhaps we should have something similar for this new benchmark method explaining why it is there? It currently says "// Ensure that we only create ZipFile.Source.Key entry per unique zip file" which doesn't seem accurate considering that this is a benchmark method and doesn't have any verifications of Key counts. 

I'll read up a bit more about jmh tomorrow because I'm unsure how multiple `@Benchmark` methods within a single class are run and what all those annotations mean when it comes to state management and such. But you don't have to wait for my approval of that part. If anyone else approves this in the meantime, please go ahead with your integration.

**Review by jaikiran — APPROVED** [review]

**Review by LanceAndersen — APPROVED** [review]
I think the latest changes are fine Sean

**Review by jaikiran — APPROVED** [review]

**Comment by coffeys** [comment]
Thanks for the reviews - test-repeat=50 passed on mach5 for product/debug platforms Jai. All looks good.

/integrate

**Comment by eirbjo** [comment]
I noticed the ZipSourceCache test fail on GHA on windows-x64:

```
FAILED     ZipSourceCache::testKeySourceMapping 'testKeySourceMapping()'
java.nio.file.FileSystemException: 1702471080605-bug8317678.zip: The process cannot access the file because it is being used by another process
	at java.base/sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:92)
	at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:103)
	at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:108)
	at java.base/sun.nio.fs.WindowsFileSystemProvider.implDelete(WindowsFileSystemProvider.java:273)
	at java.base/sun.nio.fs.AbstractFileSystemProvider.deleteIfExists(AbstractFileSystemProvider.java:109)
	at java.base/java.nio.file.Files.deleteIfExists(Files.java:1191)
	at ZipSourceCache.cleanup(ZipSourceCache.java:65)
```

On a re-run of the jobs, the test ran green again on the same source. So this seems timing related.

**Comment by jaikiran** [comment]
Hello Eirik, I have opened https://bugs.openjdk.org/browse/JDK-8322078 to track this failure.

**Pull request closed without merging**
